pallets/reward-distribution/src/lib.rs

```diff
@@ -94,10 +161,213 @@ pub mod pallet {
+				return Err(Error::<T>::NoRewardsForAccount.into())
+			}
+
+			//transfer rewards
+			Self::transfer_reward(reward_currency_id, caller_rewards, caller)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn transfer_reward(
+		reward_currency_id: CurrencyId<T>,
+		reward: BalanceOf<T>,
+		beneficiary: T::AccountId,
+	) -> DispatchResult {
+		let native_currency_id = T::GetNativeCurrencyId::get();
+		if reward_currency_id == native_currency_id {
+			let available_native_funds = T::Balances::free_balance(&Self::fee_pool_account_id());
+
+			match reward.checked_sub(&available_native_funds) {
+				None => {
+					// Pay the whole reward from the fee pool
+					let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+					return amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+				},
+				Some(remaining) => {
+					//check if the to-be-minted amount is consistent
+					//with reward tracking
+					let liability = NativeLiability::<T>::get()
+						.ok_or(Error::<T>::NotEnoughRewardsRegistered)?;
+
+					if liability < remaining {
+						return Err(Error::<T>::NotEnoughRewardsRegistered.into())
+					}
+
+					// Use the available funds from the fee pool
+					let available_amount: currency::Amount<T> =
+						Amount::new(available_native_funds, reward_currency_id);
+					available_amount.transfer(&Self::fee_pool_account_id(), &beneficiary)?;
+					// Mint the rest
+					T::Balances::deposit_creating(&beneficiary, remaining);
+
+					NativeLiability::<T>::set(Some(liability - remaining));
+				},
+			}
+			Ok(())
+		} else {
+			//we need no checking of available funds, since the fee pool will ALWAYS have enough
+			// collected fees of the wrapped currencies
+			let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+			amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+		}
+	}
+
+	pub fn execute_on_init(_height: T::BlockNumber) {
+		//get reward per block
+		let reward_per_block = match RewardPerBlock::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("Reward per block is None");
+				return
+			},
+		};
+
+		if let Err(_) = ext::security::ensure_parachain_status_running::<T>() {
+			return
+		}
+		let rewards_adapted_at = match RewardsAdaptedAt::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("RewardsAdaptedAt is None");
+				return
+			},
+		};
+
+		let mut reward_this_block = reward_per_block;
+
+		//update the reward per block if decay interval passed
+		if let Ok(expired) = ext::security::parachain_block_expired::<T>(
+			rewards_adapted_at,
+			T::DecayInterval::get() - T::BlockNumber::one(),
+		) {
+			if expired {
+				let decay_rate = T::DecayRate::get();
+				reward_this_block = (Perquintill::one() - decay_rate).mul_floor(reward_per_block);
+				RewardPerBlock::<T>::set(Some(reward_this_block));
+				let active_block = ext::security::get_active_block::<T>();
+				RewardsAdaptedAt::<T>::set(Some(active_block));
+			}
```

Comment:
The only problem I see with the new code is that we will not differentiate between an Err(_) an Ok(false), unless we remove the log bellow then it won't be a problem.

Reply:
Do we need to differentiate between them? I mean.. you have a point actually that this might not be ideal in all cases. But looking at the [parachain_block_expired](https://github.com/pendulum-chain/spacewalk/blob/a43141b920b0292ad5ec90f5c5df5729bc92b5ba/pallets/security/src/lib.rs#L256) function, the only problem we could have is an Overflow if we use really large values so maybe it's okay if we miss this case. What do you think? If you think it's critical, we can also try to keep it such that we keep the log

Reply:
Let's go with not differentiate them. I hadn't looked at the function but as you say the error is not that interesting.